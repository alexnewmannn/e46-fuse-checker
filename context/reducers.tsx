type activeFusePayload = {
  activeFuse: number;
  hideInactiveFuses?: boolean;
};

type setActiveFuseActionType = {
  type: string;
  payload: activeFusePayload;
};

const reducers = (state: Object, action: setActiveFuseActionType): object => {
  switch (action.type) {
    case 'RESET_ACTIVE_FUSE':
      return {
        ...state,
        ...action.payload,
        errors: [],
      };

    case 'SET_ACTIVE_FUSE':
      return {
        ...state,
        ...action.payload,
        errors: [],
      };

    case 'SET_EMPTY_FUSE':
      return {
        ...state,
        ...action.payload,
        errors: [
          {
            message: `No fuses were found for fuse number ${action.payload.activeFuse}. Note: fuses 1-4, 16-21 are empty as standard and 71 is the highest fuse.`,
          },
        ],
      };
    default:
      throw new Error(`Unknown action: ${action.type}`);
  }
};

export default reducers;
