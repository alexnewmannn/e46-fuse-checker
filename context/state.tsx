import React, { createContext, useContext, useReducer } from 'react';
import reducers from './reducers';

type errorType = {
  message: string;
  status?: number;
};

interface InitialState {
  errors: errorType[];
  hideInactiveFuses: boolean;
}

const initialState: InitialState = {
  hideInactiveFuses: false,
  errors: [],
};

interface Context {
  state: {};
  dispatch: React.Dispatch<any>;
}

const AppStateContext = createContext<Context>({
  state: initialState,
  dispatch: () => null,
});

export const AppWrapper: React.FC = ({ children }) => {
  const [state, dispatch] = useReducer(reducers, initialState);

  return (
    <AppStateContext.Provider value={{ dispatch, state }}>
      {children}
    </AppStateContext.Provider>
  );
};

export const useAppStateContext: Function = () => useContext(AppStateContext);
