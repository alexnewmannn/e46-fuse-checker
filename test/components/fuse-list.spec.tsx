import React from 'react';
import { axe } from 'jest-axe';
import FuseList from '../../components/fuse-list';
import { render, screen, fireEvent } from '../test-utils';
import fusesData from '../../data/fuses.json';
import { AppWrapper } from '../../context/state';

jest.mock('../../data/fuses.json');

describe('FuseList', () => {
  it('should match the snapshots', () => {
    const { asFragment, debug } = render(<FuseList />, {});
    expect(asFragment()).toMatchSnapshot();
  });

  it('should generate the correct amount of elements', () => {
    const { container } = render(<FuseList />, {});
    expect(container.querySelectorAll('dl').length).toBe(1);
    expect(container.querySelectorAll('dd').length).toBe(11);
    expect(container.querySelectorAll('dt').length).toBe(11);

    // Make sure the amount of items in the dataset matches the amount of generated elements
    expect(fusesData.length).toBe(11);
  });

  it('should have no accessibility violations', async () => {
    const { container } = render(<FuseList />, {});
    const results = await axe(container);

    expect(results).toHaveNoViolations();
  });

  describe('user interactions', () => {
    it('it should change colour of the fuses the user has clicked', async () => {
      render(
        <AppWrapper>
          <FuseList />
        </AppWrapper>
      );

      const buttons = await screen.getAllByText('52');

      expect(buttons.length).toBe(2);

      expect(buttons[0]).toHaveStyle('background: #e6e6e7');
      expect(buttons[1]).toHaveStyle('background: #e6e6e7');

      fireEvent.click(buttons[0]);

      expect(buttons[0]).toHaveStyle('background: #33a0d1');
      expect(buttons[1]).toHaveStyle('background: #33a0d1');
    });
  });
});
