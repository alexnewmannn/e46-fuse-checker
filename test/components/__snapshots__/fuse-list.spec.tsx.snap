// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FuseList should match the snapshots 1`] = `
<DocumentFragment>
  <dl
    class="fuse-list__Dl-sc-1gw2dho-0 iQYUXI"
  >
    <div
      class="fuse-list__ItemWrapper-sc-1gw2dho-3 jwefMJ"
    >
      <dd
        class="fuse-list__Dd-sc-1gw2dho-1 dEPfZp"
      >
        ABS, ASC
      </dd>
      <dt
        class="fuse-list__Dt-sc-1gw2dho-2 iVbKya"
      >
        <button
          aria-label="Fuse number 12 used by ABS, ASC"
          class="fuse-item__Fuse-sc-ubvqww-0 iQKMCk"
          type="button"
        >
          12
        </button>
        <button
          aria-label="Fuse number 33 used by ABS, ASC"
          class="fuse-item__Fuse-sc-ubvqww-0 iQKMCk"
          type="button"
        >
          33
        </button>
        <button
          aria-label="Fuse number 42 used by ABS, ASC"
          class="fuse-item__Fuse-sc-ubvqww-0 iQKMCk"
          type="button"
        >
          42
        </button>
        <button
          aria-label="Fuse number 53 used by ABS, ASC"
          class="fuse-item__Fuse-sc-ubvqww-0 iQKMCk"
          type="button"
        >
          53
        </button>
        <button
          aria-label="Fuse number 56 used by ABS, ASC"
          class="fuse-item__Fuse-sc-ubvqww-0 iQKMCk"
          type="button"
        >
          56
        </button>
      </dt>
    </div>
    <div
      class="fuse-list__ItemWrapper-sc-1gw2dho-3 jwefMJ"
    >
      <dd
        class="fuse-list__Dd-sc-1gw2dho-1 dEPfZp"
      >
        Adjustment driver seat
      </dd>
      <dt
        class="fuse-list__Dt-sc-1gw2dho-2 iVbKya"
      >
        <button
          aria-label="Fuse number 65 used by Adjustment driver seat"
          class="fuse-item__Fuse-sc-ubvqww-0 iQKMCk"
          type="button"
        >
          65
        </button>
      </dt>
    </div>
    <div
      class="fuse-list__ItemWrapper-sc-1gw2dho-3 jwefMJ"
    >
      <dd
        class="fuse-list__Dd-sc-1gw2dho-1 dEPfZp"
      >
        Adjustment passenger seat
      </dd>
      <dt
        class="fuse-list__Dt-sc-1gw2dho-2 iVbKya"
      >
        <button
          aria-label="Fuse number 70 used by Adjustment passenger seat"
          class="fuse-item__Fuse-sc-ubvqww-0 iQKMCk"
          type="button"
        >
          70
        </button>
      </dt>
    </div>
    <div
      class="fuse-list__ItemWrapper-sc-1gw2dho-3 jwefMJ"
    >
      <dd
        class="fuse-list__Dd-sc-1gw2dho-1 dEPfZp"
      >
        Airbag
      </dd>
      <dt
        class="fuse-list__Dt-sc-1gw2dho-2 iVbKya"
      >
        <button
          aria-label="Fuse number 11 used by Airbag"
          class="fuse-item__Fuse-sc-ubvqww-0 iQKMCk"
          type="button"
        >
          11
        </button>
      </dt>
    </div>
    <div
      class="fuse-list__ItemWrapper-sc-1gw2dho-3 jwefMJ"
    >
      <dd
        class="fuse-list__Dd-sc-1gw2dho-1 dEPfZp"
      >
        Air conditioner
      </dd>
      <dt
        class="fuse-list__Dt-sc-1gw2dho-2 iVbKya"
      >
        <button
          aria-label="Fuse number 28 used by Air conditioner"
          class="fuse-item__Fuse-sc-ubvqww-0 iQKMCk"
          type="button"
        >
          28
        </button>
        <button
          aria-label="Fuse number 62 used by Air conditioner"
          class="fuse-item__Fuse-sc-ubvqww-0 iQKMCk"
          type="button"
        >
          62
        </button>
        <button
          aria-label="Fuse number 63 used by Air conditioner"
          class="fuse-item__Fuse-sc-ubvqww-0 iQKMCk"
          type="button"
        >
          63
        </button>
      </dt>
    </div>
    <div
      class="fuse-list__ItemWrapper-sc-1gw2dho-3 jwefMJ"
    >
      <dd
        class="fuse-list__Dd-sc-1gw2dho-1 dEPfZp"
      >
        Automatic fearbox diesel
      </dd>
      <dt
        class="fuse-list__Dt-sc-1gw2dho-2 iVbKya"
      >
        <button
          aria-label="Fuse number 64 used by Automatic fearbox diesel"
          class="fuse-item__Fuse-sc-ubvqww-0 iQKMCk"
          type="button"
        >
          64
        </button>
      </dt>
    </div>
    <div
      class="fuse-list__ItemWrapper-sc-1gw2dho-3 jwefMJ"
    >
      <dd
        class="fuse-list__Dd-sc-1gw2dho-1 dEPfZp"
      >
        Blower
      </dd>
      <dt
        class="fuse-list__Dt-sc-1gw2dho-2 iVbKya"
      >
        <button
          aria-label="Fuse number 28 used by Blower"
          class="fuse-item__Fuse-sc-ubvqww-0 iQKMCk"
          type="button"
        >
          28
        </button>
        <button
          aria-label="Fuse number 50 used by Blower"
          class="fuse-item__Fuse-sc-ubvqww-0 iQKMCk"
          type="button"
        >
          50
        </button>
      </dt>
    </div>
    <div
      class="fuse-list__ItemWrapper-sc-1gw2dho-3 jwefMJ"
    >
      <dd
        class="fuse-list__Dd-sc-1gw2dho-1 dEPfZp"
      >
        Brake light
      </dd>
      <dt
        class="fuse-list__Dt-sc-1gw2dho-2 iVbKya"
      >
        <button
          aria-label="Fuse number 9 used by Brake light"
          class="fuse-item__Fuse-sc-ubvqww-0 iQKMCk"
          type="button"
        >
          9
        </button>
      </dt>
    </div>
    <div
      class="fuse-list__ItemWrapper-sc-1gw2dho-3 jwefMJ"
    >
      <dd
        class="fuse-list__Dd-sc-1gw2dho-1 dEPfZp"
      >
        Central locking system
      </dd>
      <dt
        class="fuse-list__Dt-sc-1gw2dho-2 iVbKya"
      >
        <button
          aria-label="Fuse number 49 used by Central locking system"
          class="fuse-item__Fuse-sc-ubvqww-0 iQKMCk"
          type="button"
        >
          49
        </button>
        <button
          aria-label="Fuse number 52 used by Central locking system"
          class="fuse-item__Fuse-sc-ubvqww-0 iQKMCk"
          type="button"
        >
          52
        </button>
        <button
          aria-label="Fuse number 60 used by Central locking system"
          class="fuse-item__Fuse-sc-ubvqww-0 iQKMCk"
          type="button"
        >
          60
        </button>
      </dt>
    </div>
    <div
      class="fuse-list__ItemWrapper-sc-1gw2dho-3 jwefMJ"
    >
      <dd
        class="fuse-list__Dd-sc-1gw2dho-1 dEPfZp"
      >
        Central locking system (only touring)
      </dd>
      <dt
        class="fuse-list__Dt-sc-1gw2dho-2 iVbKya"
      >
        <button
          aria-label="Fuse number 49 used by Central locking system (only touring)"
          class="fuse-item__Fuse-sc-ubvqww-0 iQKMCk"
          type="button"
        >
          49
        </button>
        <button
          aria-label="Fuse number 52 used by Central locking system (only touring)"
          class="fuse-item__Fuse-sc-ubvqww-0 iQKMCk"
          type="button"
        >
          52
        </button>
        <button
          aria-label="Fuse number 58 used by Central locking system (only touring)"
          class="fuse-item__Fuse-sc-ubvqww-0 iQKMCk"
          type="button"
        >
          58
        </button>
        <button
          aria-label="Fuse number 60 used by Central locking system (only touring)"
          class="fuse-item__Fuse-sc-ubvqww-0 iQKMCk"
          type="button"
        >
          60
        </button>
      </dt>
    </div>
    <div
      class="fuse-list__ItemWrapper-sc-1gw2dho-3 jwefMJ"
    >
      <dd
        class="fuse-list__Dd-sc-1gw2dho-1 dEPfZp"
      >
        Cigarette lighter
      </dd>
      <dt
        class="fuse-list__Dt-sc-1gw2dho-2 iVbKya"
      >
        <button
          aria-label="Fuse number 47 used by Cigarette lighter"
          class="fuse-item__Fuse-sc-ubvqww-0 iQKMCk"
          type="button"
        >
          47
        </button>
      </dt>
    </div>
  </dl>
</DocumentFragment>
`;
